src/java_bytecode/mz_zip_archive.cpp

```diff
@@ -0,0 +1,68 @@
+#include "mz_zip_archive.h"
+#include <stdexcept>
+#include <string>
+#include <vector>
+#define _LARGEFILE64_SOURCE 1
+#include <miniz/miniz.h>
+
+// Original struct is an anonymous struct with a typedef, This is
+// required to remove internals from the header file
+struct mz_zip_archive_statet : public mz_zip_archive { };
+
+mz_zip_archivet::mz_zip_archivet(const std::string& filename):
+  m_state(new mz_zip_archive_statet { })
+{
+  static_assert(sizeof(mz_uint)<=sizeof(size_t),
+                "size_t cannot store mz_zip file ids, choose a larger type");
+  if(MZ_TRUE!=mz_zip_reader_init_file(m_state.get(), filename.data(), 0))
+    throw std::runtime_error("MZT: Could not load a file: "+filename);
+}
+
+mz_zip_archivet::mz_zip_archivet(mz_zip_archivet &&other):
+  m_state(std::move(other.m_state)) { }
+
+mz_zip_archivet::~mz_zip_archivet()
+{
+  if(m_state)
+    mz_zip_reader_end(m_state.get());
+}
+
+mz_zip_archivet& mz_zip_archivet::operator=(mz_zip_archivet &&other)
+{
+  if(m_state)
+    mz_zip_reader_end(m_state.get());
+  m_state=std::move(other.m_state);
+  return *this;
+}
+
+size_t mz_zip_archivet::get_num_files()
+{
+  return mz_zip_reader_get_num_files(m_state.get());
+}
+
+std::string mz_zip_archivet::get_filename(const size_t file_id)
+{
+  const auto id=static_cast<mz_uint>(file_id);
+  std::vector<char> buffer(
+    mz_zip_reader_get_filename(m_state.get(), id, nullptr, 0));
+  mz_zip_reader_get_filename(m_state.get(), id, buffer.data(), buffer.size());
+  // Cannot construct from iterators. Junk returned after last '\0'
+  return buffer.data();
```

Comment:
Sure, but the miniz.c file will presumably get updated in future, and failure to properly terminate a buffer in rare circumstances would hardly be a big surprise.